### PR TITLE
feat(metrics): Add prometheus metrics

### PR DESCRIPTION
**Pull request by someone-stole-my-name**
This allows users to get insights about what is the handler doing without having to manually parse logs.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -54,6 +59,22 @@ func Initialize() error {
...
+		if err != nil {
+			return fmt.Errorf("invalid value for '%s': %s", EnvMetricsPort, err)
+		}
+		cfg.MetricsPort = port
+	}
+
+	if metrics := strings.ToLower(os.Getenv(EnvMetrics)); len(metrics) != 0 {
+		cfg.Metrics = true
+	}
+
```
Comment (TwiN) [inline config/config.go:77]:
```suggestion
	if metricsPort := os.Getenv(EnvMetricsPort); len(metricsPort) == 0 {
		log.Printf("Environment variable '%s' not specified, defaulting to 8080", EnvMetricsPort)
		cfg.MetricsPort = 8080
	} else {
		port, err := strconv.Atoi(metricsPort)
		if err != nil {
			return fmt.Errorf("invalid value for '%s': %s", EnvMetricsPort, err)
		}
		cfg.MetricsPort = port
	}
	if metrics := strings.ToLower(os.Getenv(EnvMetrics)); len(metrics) != 0 {
		cfg.Metrics = true
	}
```
Just to clear up the empty lines

File: metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		}, []string{"node_group"}),
+		DrainedNodes: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "drained_nodes_total",
+			Help:      "The total number of drained nodes",
+		}, []string{"node_group"}),
+		Errors: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "errors",
+			Help:      "The total errors",
```
Comment (TwiN) [inline metrics/metrics.go:69]:
```suggestion
			Help:      "The total number of errors",
```

File: metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		}),
+	}
+
+	m.register()
+
```
Comment (TwiN) [inline metrics/metrics.go:74]:
```suggestion
	m.register()
```

**Review by TwiN — CHANGES_REQUESTED** [review]
Some small suggestions, but aside from that, excellent work!
Thank you for contributing :)

**Comment by someone-stole-my-name** [comment]
I've made all changes and squashed.

File: README.md [hunk 4]
```diff
@@ -51,7 +51,22 @@ Therefore, this application will not run into any issues if it is restarted, res
 | ENVIRONMENT              | If set to `dev`, will try to create the Kubernetes client using your local kubeconfig. Any other values will use the in-cluster configuration | no | `""` |
 | EXECUTION_INTERVAL | Duration to sleep between each execution in seconds | no | `20` |
 | EXECUTION_TIMEOUT | Maximum execution duration before timing out in seconds | no | `900` |
-
+| METRICS_PORT             | Port to bind metrics server to | no | `8080` |
+| METRICS                  | Expose metrics in Promtheus format at `:${METRICS_PORT}/metrics` | no | `""` | 
+
+## Metrics
+
+All the metrics are prefixed with `rolling_update_handler`.
```
Comment (TwiN) [inline README.md:59]: Can you just prefix the metrics in the table with that instead? It'll make searching using GitHub easier

File: main.go [hunk 5]
```diff
@@ -72,11 +77,13 @@ func run(ec2Service ec2iface.EC2API, autoScalingService autoscalingiface.AutoSca
...
 	}
 	if err != nil {
+		metrics.Server.Errors.Inc()
 		return errors.New("unable to describe AutoScalingGroups: " + err.Error())
 	}
 	if cfg.Debug {
 		log.Println("Described AutoScalingGroups successfully")
 	}
+	metrics.Server.NodeGroups.WithLabelValues().Set(float64(len(autoScalingGroups)))
 	return HandleRollingUpgrade(kubernetesClient, ec2Service, autoScalingService, autoScalingGroups)
```
Comment (TwiN) [inline main.go:87]: Can you move `metrics.Server.NodeGroups.WithLabelValues().Set(float64(len(autoScalingGroups)))` inside `HandleRollingUpgrade`?

File: main.go [hunk 6]
```diff
@@ -58,6 +62,7 @@ func run(ec2Service ec2iface.EC2API, autoScalingService autoscalingiface.AutoSca
 	cfg := config.Get()
 	client, err := k8s.CreateClientSet()
 	if err != nil {
+		metrics.Server.Errors.Inc()
 		return errors.New("unable to create Kubernetes client: " + err.Error())
 	}
```
Comment (TwiN) [inline main.go:67]: Can you move this `metrics.Server.Errors.Inc()` to the parent function instead? That way you won't have to do it both here, on line 80 and on line 105 of the same file.

**Review by TwiN — CHANGES_REQUESTED** [review]
I'm so sorry for the delay, I've been very busy 😔

**Review by TwiN — APPROVED** [review]

**Comment by TwiN** [comment]
Awesome work, thank you for the contribution!

**Comment by TwiN** [comment]
Released in [v1.4.0](https://github.com/TwiN/aws-eks-asg-rolling-update-handler/releases/tag/v1.4.0)

**Pull request merged**